src/group_theory/quaternion_group.lean

```diff
@@ -0,0 +1,289 @@
+`quaternion_group 1` is isomorphic to a cyclic group of order `4`.
+
+## References
+
+https://en.wikipedia.org/wiki/Dicyclic_group
+https://en.wikipedia.org/wiki/Quaternion_group
+
+## TODO
+
+Show that `quaternion_group 2 ≃* units (quaternion ℤ)`.
+
+-/
+
+/--
+The (generalised) quaternion group `quaternion_group n` of order `4n`. It can be defined by the
+presentation `⟨a, x | a²ⁿ = 1, x² = aⁿ, x⁻¹ax=a⁻¹⟩`. We write `a i` for `aⁱ` and `xa i` for `xaⁱ`.
+-/
+@[derive decidable_eq]
+inductive quaternion_group (n : ℕ) : Type
+| a : zmod (2 * n) → quaternion_group
+| xa : zmod (2 * n) → quaternion_group
+
+namespace quaternion_group
+
+variables {n : ℕ}
+
+/--
+Multiplication of the dihedral group.
+-/
+private def mul : quaternion_group n → quaternion_group n → quaternion_group n
+| (a i) (a j) := a (i + j)
+| (a i) (xa j) := xa (j - i)
+| (xa i) (a j) := xa (i + j)
+| (xa i) (xa j) := a (n + j - i)
+
+/--
+The identity `1` is given by `aⁱ`.
+-/
+private def one : quaternion_group n := a 0
+
+instance : inhabited (quaternion_group n) := ⟨one⟩
+
+/--
+The inverse of a an element of the dihedral group.
+-/
+private def inv : quaternion_group n → quaternion_group n
+| (a i) := a (-i)
+| (xa i) := xa (n + i)
+
+/--
+The group structure on `quaternion_group n`.
+-/
+instance : group (quaternion_group n) :=
+{ mul := mul,
+  mul_assoc :=
+  begin
+    rintros (i | i) (j | j) (k | k);
+    simp only [mul];
+    ring,
+    simp only [add_left_inj, add_right_inj],
+    calc -(n : zmod (2 * n)) = 0 - (n : zmod (2 * n)) : by rw zero_sub
+      ... = (2 * n : zmod (2 * n)) - (n : zmod (2 * n) ) :
+        by { rw ← zmod.nat_cast_self (2 * n), norm_cast }
+      ... = (n : zmod (2 * n)) : by ring
+  end,
+  one := one,
+  one_mul :=
+  begin
+    rintros (i | i),
+    exact congr_arg a (zero_add i),
+    exact congr_arg xa (sub_zero i),
+  end,
+  mul_one := begin
+    rintros (i | i),
+    exact congr_arg a (add_zero i),
+    exact congr_arg xa (add_zero i),
+  end,
+  inv := inv,
+  mul_left_inv := begin
+    rintros (i | i),
+    exact congr_arg a (neg_add_self i),
+    exact congr_arg a (sub_self (n + i)),
+  end }
+
+variable {n}
+
+@[simp] lemma a_mul_a (i j : zmod (2 * n)) : a i * a j = a (i + j) := rfl
+@[simp] lemma a_mul_xa (i j : zmod (2 * n)) : a i * xa j = xa (j - i) := rfl
+@[simp] lemma xa_mul_a (i j : zmod (2 * n)) : xa i * a j = xa (i + j) := rfl
+@[simp] lemma xa_mul_xa (i j : zmod (2 * n)) : xa i * xa j = a (n + j - i) := rfl
+
+lemma one_def : (1 : quaternion_group n) = a 0 := rfl
+
+private def fintype_helper : (zmod (2 * n) ⊕ zmod (2 * n)) ≃ quaternion_group n :=
+{ inv_fun := λ i, match i with
+                 | (a j) := sum.inl j
+                 | (xa j) := sum.inr j
+                 end,
+  to_fun := λ i, match i with
+                 | (sum.inl j) := a j
+                 | (sum.inr j) := xa j
+                 end,
+  left_inv := by rintro (x | x); refl,
+  right_inv := by rintro (x | x); refl }
```

Comment:
No criticism intended here, but we really need someone to write a `fintype` derive handler. Humans should not be involved here. :-)

Reply:
We have a fintype derive handler, (but I've never used it before so I don't know if it helps here): https://leanprover-community.github.io/mathlib_docs/tactic/derive_fintype.html

Reply:
Sticking `derive fintype` on the definition results in 
```failed to synthesize type class instance for
n : ℕ
⊢ fintype (Σ' (ᾰ : zmod (2 * n)), unit)```

Reply:
That's because it's not a `fintype` when `n = 0` as `zmod n` is `ℤ`, so I don't think you can derive `fintype` for it.

Reply:
That's because it's not a `fintype` when `n = 0`.

Reply:
In `dihedral_group`, I decided not to separate out `dihedral n` when `0 < n` and `infinite_dihedral` like the original PR by Neil Strickland, since `infinite_dihedral` would be defeq to `dihedral 0`. What this does mean though is that we can't derive `fintype` for `dihedral` as when `n = 0`. Same goes for `quaternion_group` here.

Reply:
Maybe it would be a good idea to make `derive_fintype` accept additional conditions like here `0 < n`.

Reply:
Does `[derive [\lam T [fact (0 < n)], fintype T]]` work?

Reply:
@eric-wieser No, it seems to get confused of what the `n` is in that fact.